core/types/calls.cc

```diff
@@ -1114,7 +1147,11 @@ class T_Generic_squareBrackets : public IntrinsicMethod {
             }
         }
 
-        res.returnType = make_type<MetaType>(make_type<AppliedType>(attachedClass, move(targs)));
+        if (validBounds) {
```

Comment:
https://github.com/sorbet/sorbet/pull/1392/files#r307538043 
will allow to get rid from this boolean